### PR TITLE
Htmlhelper unittests

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Partial fix for issue #453 
Will send another PR for the remaining extensions.
@dougbu

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        }
+
+        [Fact]
+        public void CheckboxHelpers_UsesSpecifiedIsChecked()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var checkboxResult = helper.CheckBox("Property1", true);
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: Consider named param for this (considering that's the whole purpose of this test).

**Comment by Eilon** [comment]
BTW are these tests ported from the old code, or did you write brand new ones?

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+using Microsoft.AspNet.Mvc.Rendering;
+using Microsoft.AspNet.Mvc.TestCommon;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core
+{
+    /// <summary>
+    /// Test the <see cref="HtmlHelperInputExtensions" /> class.
+    /// </summary>
+    public class HtmlHelperInputExtensionsTest
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: Given the other test classes in the folder, this grouping is quite odd. Why aren't you adding `CheckBoxExtensionsTest`, `HiddenExtensionsTest` and so on?

**Comment by ajaybhargavb** [comment]
I wrote these tests. These tests covered the untested extension methods in `HtmlHelperInputExtensions`

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: When we discussed yesterday, I thought you said its better to add a corresponding test class for every extension class.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var checkboxResult = helper.CheckBox("Property1");
+
+            // Assert
+            Assert.Equal(
+                "<input id=\"HtmlEncode[[Property1]]\" name=\"HtmlEncode[[Property1]]\" type=\"HtmlEncode[[checkbox]]\" value=\"HtmlEncode[[true]]\" />" +
+                "<input name=\"HtmlEncode[[Property1]]\" type=\"HtmlEncode[[hidden]]\" value=\"HtmlEncode[[false]]\" />",
+                HtmlContentUtilities.HtmlContentToString(checkboxResult));
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: Not a great approach for testing extension methods. As we discussed offline, we have some tests in this folder following the pattern you're using here. But it's better to clearly separate tests of the actual `IHtmlHelper` and `IHtmlHelper<T>` methods from tests of the extensions.

To do that, suggest you either mock `IHtmlGenerator` (see `HtmlHelperFormExtensionsTest`) or `IHtmlHelper` / `IHtmlHelper<T>` (see `HtmlHelperPartialExtensionsTest`). I prefer the second because it's more straightforward.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: Hmm, meant "extension _method_". See existing `HtmlHelperHiddenTest` and `HtmlHelperPasswordTest` classes.

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs (hunk 3 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: If you mock the HTML generator, what would the output be here? I think it's nice in the unit tests to see what the rendered HTML will look like.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: @Eilon I agree in general but not for these tests. As-is, they aren't unit tests and are instead testing `DefaultHtmlGenerator`, `HtmlHelper`, `HtmlHelper<T>` as well as `HtmlHelperInputExtensions`. So I'd like to see lots of checks of the rendered HTML in unit tests of `HtmlHelper` and `HtmlHelper<T>`.

Could take that to an extreme and mock `IHtmlGenerator` when testing `HtmlHelper`. But we don't have many tests like that today and `IHtmlGenerator` is something of an implementation detail for HTML and tag helpers.

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: I think regardless of philosophy, it's a strong value-add to have tests that cover the exact HTML output by various html helper apis.

In my experience most changes that impact the HTML output are only caught in functional tests and it's hard to investigate.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: I'm agreeing on the value of testing to confirm the generated HTML. But I would rather that be done in tests that call `IHtmlHelper` methods directly. Here it just gets redundant.

Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperInputExtensionsTest.cs]: The tests will be too mechanical to provide any real value. The tests as they are now will catch real bugs. Let's stick with this approach.

**Comment by ajaybhargavb** [comment]
Updated.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs [hunk 4]
```diff
@@ -649,6 +649,54 @@ public void CheckBoxFor_WithComplexExpressions_DoesNotUseValuesFromViewDataDicti
             Assert.Equal(expected, HtmlContentUtilities.HtmlContentToString(html));
         }
 
+        [Fact]
+        public void CheckboxHelpers_UsesSpecifiedExpression()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: Careful of that 's' -- `CheckboxHelpers_Use...` or `CheckboxHelper_Uses...`.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: ... throughout the PR.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperHiddenTest.cs [hunk 5]
```diff
@@ -835,6 +835,40 @@ public void HiddenFor_DoesNotUseAttributeValue()
...
 
+        [Fact]
+        public void HiddenHelpers_UsesSpecifiedExpression()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var hiddenResult = helper.Hidden("Property1");
+            var hiddenForResult = helper.HiddenFor(m => m.Property1);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperHiddenTest.cs]: Not a fan of doing two independent tests in the same `[Fact]` though we have some ugly examples in these HTML helper tests.

Split into separate methods. Or (perhaps) pass `Func<IHtmlHelper<ForgetTheType>, IHtmlContent>` into a `[Theory]` (since the content matches here).

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperLabelExtensionsTest.cs [hunk 6]
```diff
@@ -252,6 +252,38 @@ public void LabelFor_DisplaysVariableName()
...
+        [Fact]
+        public void LabelHelpers_UsesSpecifiedLabelText()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var labelResult = helper.Label("Property1", labelText: "Hello");
+            var labelForResult = helper.LabelFor(m => m.Property1, labelText: "Hello");
+            var labelForModelResult = helper.LabelForModel(labelText: "Hello");
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperLabelExtensionsTest.cs]: Split up.

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperLabelExtensionsTest.cs]: Also this last one is testing something different. Other calls get `<label>` for the `Property1` method but this gets the `<label>` for the entire model. The second is a rather special case since, in many cases, the top-level model cannot be used in HTML helpers.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperLabelExtensionsTest.cs [hunk 7]
```diff
@@ -252,6 +252,38 @@ public void LabelFor_DisplaysVariableName()
...
+
+        [Fact]
+        public void LabelHelpers_DisplaysSpecifiedHtmlAttributes()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var labelForResult = helper.LabelFor(m => m.Property1, htmlAttributes: new { attr="value" });
+            var labelForModelResult = helper.LabelForModel(labelText: "Hello", htmlAttributes: new { attr="value" });
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperLabelExtensionsTest.cs]: Split these two different scenarios. Again, `LabelFor(m => m)` is the call that's analogous to `LabelForModel()` but both are special cases unless you're in a partial, `EditorFor()`, or similar.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs [hunk 8]
```diff
@@ -345,6 +345,40 @@ public static TheoryData PasswordFor_WithComplexExpressionsData
...
 
+        [Fact]
+        public void PasswordHelpers_UsesSpecifiedExpression()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var passwordResult = helper.Password("Property1");
+            var passwordForResult = helper.PasswordFor(m => m.Property1);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs]: `[Theory]` or two `[Fact]`s. (`[Theory]` would be :ok: here because expressions work out to the same thing and output is identical.)

Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs]: ... same for remaining tests. I don't see more that test the `...ForModel()` extensions.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+using Microsoft.AspNet.Mvc.Rendering;
+using Microsoft.AspNet.Mvc.TestCommon;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core
+{
+    /// <summary>
+    /// Test the RadioButton extensions in <see cref="HtmlHelperInputExtensions" /> class.
+    /// </summary>
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs]: Fine but don't think doc comments are at all required in a regular test class.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    public class HtmlHelperRadioButtonExtensionsTest
+    {
+        [Fact]
+        public void RadioButtonHelpers_UsesSpecifiedExpression()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs]: This test doesn't demonstrate what you've got on the tin. Here, you're showing that `checked` isn't generated when `Property1` is something (anything) other than `"myvalue"`. Also need a test showing that `checked` _is_ generated when `Property1 == "myvalue"`. And it's also worth confirming whether the conditions are case-sensitive i.e. test what happens when `Property1 == "MyValue"`.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperHiddenTest.cs [hunk 11]
```diff
@@ -835,6 +835,40 @@ public void HiddenFor_DoesNotUseAttributeValue()
             Assert.Equal(expected, HtmlContentUtilities.HtmlContentToString(result));
         }
 
+        [Fact]
+        public void HiddenHelpers_UsesSpecifiedExpression()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperHiddenTest.cs]: `Property1` needs to be non-`null` to show anything interesting here.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs [hunk 12]
```diff
@@ -649,6 +649,54 @@ public void CheckBoxFor_WithComplexExpressions_DoesNotUseValuesFromViewDataDicti
...
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var checkboxResult = helper.CheckBox("Property1");
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: Similar to `RadioButtonHelpers_UsesSpecifiedExpression()`, must demonstrate what happens when `Property1 == "false"`, `Property1 == "FALSE"`, `Property1 = "true", and `Property1 == "Not a bool"`. Last case may already be handled (see a test using `Assert.Throws<FormatException>` below).

Suspect tests exist already covering `bool` properties. Double-check please.

I believe `bool` property

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            Assert.Equal(
+                "<input id=\"HtmlEncode[[Property1]]\" name=\"HtmlEncode[[Property1]]\" type=\"HtmlEncode[[radio]]\" value=\"HtmlEncode[[myvalue]]\" />",
+                HtmlContentUtilities.HtmlContentToString(radioButtonResult));
+            Assert.Equal(
+                "<input id=\"HtmlEncode[[Property1]]\" name=\"HtmlEncode[[Property1]]\" type=\"HtmlEncode[[radio]]\" value=\"HtmlEncode[[myvalue]]\" />",
+                HtmlContentUtilities.HtmlContentToString(radioButtonForResult));
+        }
+
+        [Fact]
+        public void RadioButtonHelpers_UsesSpecifiedIsChecked()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs]: Test the opposite case as well: `Property1 == "true"` with 'isChecked: false'.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs [hunk 14]
```diff
@@ -649,6 +649,54 @@ public void CheckBoxFor_WithComplexExpressions_DoesNotUseValuesFromViewDataDicti
...
+        }
+
+        [Fact]
+        public void CheckboxHelpers_UsesSpecifiedIsChecked()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var checkboxResult = helper.CheckBox("Property1", isChecked: true);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: Test the opposite case as well: `Property1 == "true"` with 'isChecked: false'.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs [hunk 15]
```diff
@@ -345,6 +345,40 @@ public static TheoryData PasswordFor_WithComplexExpressionsData
             Assert.Equal(expected, HtmlContentUtilities.HtmlContentToString(result));
         }
 
+        [Fact]
+        public void PasswordHelpers_UsesSpecifiedExpression()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs]: No, `Password()` and `PasswordFor()` ignore the expression value, `ViewData`, `ModelState`, and so on. _Only_ the passed `htmlAttributes` or `value` arguments can results in a generated `value` attribute. (And in the `PasswordFor()` case, there's no `value` argument.)

Need to change this test name and test a non-`null` `Property1`

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextAreaExtensionsTest.cs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace Microsoft.AspNet.Mvc.Core
+{
+    /// <summary>
+    /// Test the TextArea extensions in <see cref="HtmlHelperInputExtensions" /> class.
+    /// </summary>
+    public class HtmlHelperTextAreaExtensionsTest
+    {
+        [Fact]
+        public void TextAreaHelpers_UsesSpecifiedExpression()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextAreaExtensionsTest.cs]: Need a non-`null` value to show that it's used. Check everywhere I didn't already make this comment too.

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs (hunk 12 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: There are existing checkbox tests that test all scenarios that you mentioned above (not just the `Assert.Throws` case). I'll add these cases for `RadioButton`.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs (hunk 13 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs]: Added.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs (hunk 14 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: Added.

**Comment by ajaybhargavb** [comment]
Updated.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs [hunk 17]
```diff
@@ -345,6 +345,60 @@ public static TheoryData PasswordFor_WithComplexExpressionsData
             Assert.Equal(expected, HtmlContentUtilities.HtmlContentToString(result));
         }
 
+        [Fact]
+        public void Password_UsesSpecifiedExpressionButIgnoresValue()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs]: `Password_UsesSpecifiedExpressionForNames_IgnoresExpressionValue()`. Same for similar test below.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+            // Assert
+            Assert.Equal(
+                "<input id=\"HtmlEncode[[Property1]]\" name=\"HtmlEncode[[Property1]]\" type=\"HtmlEncode[[radio]]\" value=\"HtmlEncode[[myvalue]]\" />",
+                HtmlContentUtilities.HtmlContentToString(radioButtonForResult));
+        }
+
+        [Theory]
+        [InlineData("MyValue")]
+        [InlineData("myvalue")]
+        public void RadioButton_UsesSpecifiedExpressionCheckedWhenValueMatches(string value)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperRadioButtonExtensionsTest.cs]: `RadioButtion_CheckedWhenValueMatchesSpecifiedExpression` and same for similar test below.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxExtensionsTest.cs [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+        public void TextBox_UsesSpecifiedFormat()
+        {
+            // Arrange
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(new ViewDataDictionary<TestModel>(metadataProvider));
+            helper.ViewContext.ClientValidationEnabled = false;
+            helper.ViewData.Model = new TestModel { Property1 = "propValue" };
+
+            // Act
+            var textBoxResult = helper.TextBox("Property1", value: "myvalue", format: "prefix: {0}");
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxExtensionsTest.cs]: Add a similar test with same extension method but `value: null`.

**Comment by dougbu** [comment]
Is this PR intended to be the last word on #453? If yes, must add more tests covering open items mentioned in that issue. I'm fine doing that separately (hmm, prefer it) but don't want to see the issue close immediately.

- Indexers and deeply nested lambdas for the various `*For` e.g `InputFor`, `DisplayFor` etc
- Scenarios gap compared to 5.2
- a few other methods are exposed publicly but used in other helpers (and therefore tested indirectly, or will be):
 - `@Html.Encode()`
 - `@Html.FormatValue()`
 - `@Html.GenerateIdFromName()`
 - `@Html.GetClientValidationRules()`
 - `@Html.GetFormMethodString()`
- Some details on remaining `DropDownList` and `ListBox` test gaps.  Test do not currently:
 - use non-`null` `optionLabel` and `htmlAttributes` arguments
 - confirm value sources and their priorities
 - mock an `IHtmlGenerator` and confirm how it is called
 - mock an `IHtmlHelper` and confirm how extension methods call that

My point in the second-last bullet ([here](https://github.com/aspnet/Mvc/issues/453#issuecomment-70395759)) was to write a few (not too many) tests to cover usage that isn't already handled indirectly.

"confirm value sources and their priorities" is important in general and a big part of "Scenarios gap compared to 5.2". Other than expression vs. explicit values, that isn't covered in any of the new test classes. Need to show `ModelStateDictionary` values, `ViewDataDictionary` values, and their precedence. (Or, in the case of, `Password[For]()` those values being ignored.) Have a look at tests like `HiddenUsesValuesFromModelState_OverExplicitSpecifiedValueAndPropertyValue()` in `HtmlHelperHiddenTest`.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs [hunk 20]
```diff
@@ -649,6 +649,73 @@ public void CheckBoxFor_WithComplexExpressions_DoesNotUseValuesFromViewDataDicti
...
+        }
+
+        [Fact]
+        public void Checkbox_UsesSpecifiedHtmlAttributes()
+        {
+            // Arrange
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper();
+
+            // Act
+            var checkboxResult = helper.CheckBox("Property1", htmlAttributes: new { attr = "value" });
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperCheckboxTest.cs]: Something more to track for another PR: `htmlAttributes` is another "value source". Not sure even the existing test classes cover including `checked`, `id`, `name`, `type` or `value` properties in the `htmlAttributes` object.

FYI the generated HTML might override a `name` and `value` (when the `value` argument is non-`null`) but `htmlAttributes` should "win" in general.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationMessageExtensionsTest.cs [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+using Microsoft.AspNet.Mvc.Rendering;
+using Microsoft.AspNet.Mvc.TestCommon;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Core
+{
+    /// <summary>
+    /// Test the ValidationMessage extensions in <see cref="HtmlHelperValidationExtensions" /> class.
+    /// </summary>
+    public class HtmlHelperValidationMessageExtensionsTest
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperValidationMessageExtensionsTest.cs]: For another PR: This test class covers all extension methods but only when the property is valid and client-side validation is enabled. Need to confirm what happens for all combinations of those two characteristics e.g. invalid properties cause a `<span>` to be generated even if client-side validation is disabled.

**Comment by dougbu** [comment]
:watch: just for one more test and a couple of renames.

**Comment by ajaybhargavb** [comment]
@dougbu, no. My intention was to not close the issue.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxExtensionsTest.cs (hunk 19 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/HtmlHelperTextBoxExtensionsTest.cs]: Added.

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by dougbu** [comment]
:shipit:

Please also update #453 to make it clear what work remains. Cross out the done bits in existing comments and add another comment with a "still to come" list (likely similar to my 3 comments above though I may have missed something). I don't recommend just linking to my comments above because that buries the information.
